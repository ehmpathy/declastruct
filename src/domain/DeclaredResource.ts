import { DomainObject } from 'domain-objects';

// TODO: improve on the type of "declared resource"
// eslint-disable-next-line @typescript-eslint/ban-types
export type DeclaredResource<R extends Record<string, any> = {}> =
  DomainObject<R>;
