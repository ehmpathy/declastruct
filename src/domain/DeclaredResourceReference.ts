import { DomainLiteral } from 'domain-objects';
import Joi from 'joi';

export enum DeclaredResourceReferenceKeyType {
  PRIMARY_KEY = 'PRIMARY_KEY',
  UNIQUE_KEY = 'UNIQUE_KEY',
}

/**
 * a value that a declared resource reference identifier specifies to identify the resource
 */
export class DeclaredResourceReferenceIdentifierValue<
  V extends Record<string, any>,
> extends DomainLiteral<V> {}

/**
 * an identifier for a declared resource reference
 */
export interface DeclaredResourceReferenceIdentifier<
  KT extends DeclaredResourceReferenceKeyType,
  T,
  K extends keyof T,
> {
  key: KT;
  value: Required<Pick<T, K>>;
}
export class DeclaredResourceReferenceIdentifier<
    KT extends DeclaredResourceReferenceKeyType,
    T,
    K extends keyof T,
  >
  extends DomainLiteral<DeclaredResourceReferenceIdentifier<KT, T, K>>
  implements DeclaredResourceReferenceIdentifier<KT, T, K>
{
  public static nested = { value: DeclaredResourceReferenceIdentifierValue };
}

const schema = Joi.alternatives(
  Joi.object().keys({
    _dobj: Joi.string().optional(), // may be specified if hydrating from serialized form (e.g., a cache)
    referenceOf: Joi.string().required(),
    identifiedBy: Joi.object().keys({
      _dobj: Joi.string().optional(), // may be specified if hydrating from serialized form (e.g., a cache)
      key: Joi.string()
        .valid(...Object.values(DeclaredResourceReferenceKeyType))
        .required(),
      value: Joi.object(),
    }),
  }),
  Joi.string().required(), // alternatively, may be in a string representation if we're displaying a 'grokified' version of the resource (note: this should only occur in internal declastruct usage, so its safe that we allow the type to not match the typescript type in this way)
);

/**
 * a reference to a resource
 * - allows reference by primary key (i.e., can specify just the primary)
 * - allows reference by unique key (i.e., can specify the unique traits of the entity)
 */
export interface DeclaredResourceReference<
  T,
  /**
   * the names of the primary key attributes
   */
  P extends keyof T,
  /**
   * the names of the unique key attributes
   */
  U extends keyof T,
> {
  /**
   * the class.name of the resource being referenced
   */
  referenceOf: string;

  /**
   * the mechanism used to reference
   */
  identifiedBy:
    | {
        key: DeclaredResourceReferenceKeyType.PRIMARY_KEY;
        value: Required<Pick<T, P>>;
      }
    | {
        key: DeclaredResourceReferenceKeyType.UNIQUE_KEY;
        value: Required<Pick<T, U>>;
      };
  // TODO: determine why types dont fit with the below
  //   | DeclaredResourceReferenceIdentifier<
  //       DeclaredResourceReferenceKeyType.PRIMARY_KEY,
  //       T,
  //       P
  //     >
  //   | DeclaredResourceReferenceIdentifier<
  //       DeclaredResourceReferenceKeyType.UNIQUE_KEY,
  //       T,
  //       U
  //     >;
}
export class DeclaredResourceReference<
    T,
    /**
     * the names of the primary key attributes
     */
    P extends keyof T,
    /**
     * the names of the unique key attributes
     */
    U extends keyof T,
  >
  extends DomainLiteral<DeclaredResourceReference<T, P, U>>
  implements DeclaredResourceReference<T, P, U>
{
  public static schema = schema;
  public static nested = {
    identifiedBy: DeclaredResourceReferenceIdentifier,
  };
}
