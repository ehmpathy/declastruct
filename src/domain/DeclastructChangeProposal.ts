import { DomainObject } from 'domain-objects';

import { DeclaredResource } from './DeclaredResource';

/**
 * an action that can be proposed
 */
export enum DeclastructChangeProposalAction {
  /**
   * do nothing, the resource is already in the desired state
   */
  DO_NOTHING = 'DO_NOTHING',

  /**
   * create a new resource
   */
  CREATE = 'CREATE',

  /**
   * update an existing resource
   */
  UPDATE = 'UPDATE',

  /**
   * destroy an existing resource
   */
  DESTROY = 'DESTROY',

  /**
   * replace an existing resource
   *
   * note
   * - this is a `delete` followed by a `create`
   */
  REPLACE = 'REPLACE',
}

/**
 * a change that declastruct has proposed in order to update the remote state to reflect the declared state
 */
export interface DeclastructChangeProposal<R extends DeclaredResource> {
  /**
   * the name of the class of resource this proposal is for
   */
  forResourceClassName: string;

  /**
   * the grokable identity string of the resource this proposal is for
   */
  forGrokableIdentifier: string;

  /**
   * the remote-state the resource is currently in, from which we will be changing it
   */
  fromRemoteState: R | null;

  /**
   * the desired-state the resource should be in, to which we will be changing it
   */
  toDesiredState: R | null;

  /**
   * the action we plan to execute in order to fulfil this proposed change
   */
  action: DeclastructChangeProposalAction;

  /**
   * the displayable difference made by the change
   */
  difference: string | null;
}

export class DeclastructChangeProposal<R extends DeclaredResource>
  extends DomainObject<DeclastructChangeProposal<R>>
  implements DeclastructChangeProposal<R> {}
