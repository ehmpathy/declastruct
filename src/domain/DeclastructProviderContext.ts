import type { LogMethods } from 'simple-leveled-log-methods';

import { DeclaredResource } from './DeclaredResource';
import {
  DeclastructProvider,
  DeclastructProviderAgentOptions,
} from './DeclastructProvider';

/**
 * defines the context in which declastruct is being run, accessible a provider
 *
 * specifically
 * - the options that were inititialized for the provider
 * - the resources that were defined for the provider
 *
 * note
 * - we do not expose the full declastruct context to any provider for security
 *   - i.e., ProviderA will never be able to see context related to ProviderB
 *   - otherwise, MaliciousProvider may be able to access the credentials defined for CriticalProvider, and steal them
 *   - therefore, only the information explicitly given by the user to ProviderA is accessible to ProviderA
 */
export interface DeclastructProviderContext<
  AO extends DeclastructProviderAgentOptions,
> {
  /**
   * the options instantiated for the provider itself
   */
  provider: DeclastructProvider<AO>;

  /**
   * the resources for this provider that were declared in this invocation
   */
  resources?: DeclaredResource[];

  /**
   * the log context to use
   */
  log: LogMethods;
}
