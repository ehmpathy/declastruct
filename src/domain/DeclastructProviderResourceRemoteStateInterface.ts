import { PickOne } from 'type-fns';

import { DeclaredResource } from './DeclaredResource';
import {
  DeclaredResourceClassName,
  DeclastructProviderAgentOptions,
} from './DeclastructProvider';
import { DeclastructProviderContext } from './DeclastructProviderContext';

/**
 * an interface which enables the management of the remote state of a resource
 * - capable of crud operations on the resource
 * - conforms to a standard shape for interoperability
 */
export interface DeclastructProviderResourceRemoteStateInterface<
  /**
   * the class of resource being managed by this dao
   */
  R extends DeclaredResource,
  /**
   * the shape of options given to the agent of this resource's provider
   */
  AO extends DeclastructProviderAgentOptions,
  /**
   * the names of the primary key attributes
   */
  P extends keyof R,
  /**
   * the names of the unique key attributes
   */
  U extends keyof R,
  /**
   * the shape of the resources search filter options
   */
  F extends undefined | null | Record<string, any> = undefined,
  /**
   * the shape of the resources search sort options
   */
  S extends undefined | Record<string, any | null> = undefined,
> {
  /**
   * the name of the resource this interface manages
   */
  for: DeclaredResourceClassName;

  /**
   * a method capable of finding the state of a resource of this class, by primary-key
   */
  findByPrimary: (
    /**
     * the primary key to lookup the resource by
     */
    primary: Required<Pick<R, P>>,

    /**
     * the declastruct context available to interfaces of this provider
     */
    context: DeclastructProviderContext<AO>,
  ) => Promise<Required<R> | null>;

  /**
   * a method capable of finding the state of a resource of this class, by unique-key
   */
  findByUnique: (
    /**
     * the unique key to lookup the resource by
     */
    unique: Required<Pick<R, U>>,

    /**
     * the declastruct context available to interfaces of this provider
     */
    context: DeclastructProviderContext<AO>,
  ) => Promise<Required<R> | null>;

  /**
   * a method capable of finding all instances of a resource of this class, by some filtering criteria
   *
   * note
   * - may not be supported if search is not possible
   */
  findAll?: (
    /**
     * the criteria to search with
     */
    criteria: {
      /**
       * what to filter the results to list by
       *
       * note
       * - if null is supported and specified, no filter will be applied
       */
      filter: F;

      /**
       * what to sort the results by
       *
       * note
       * - this dictates the order of results
       * - this dictates the offset key for pagination
       *
       * tip
       * - if you're seeing "never" when trying to use this, please make sure you're using the instantiated type of the resource, not the generic type
       */
      sort: PickOne<{
        /**
         * descending sort
         */
        until: S;

        /**
         * ascending sort
         */
        since: S;
      }>;

      /**
       * how many results to return
       *
       * note
       * - combine this with the sort option to paginate results
       */
      limit: number;
    },
    /**
     * the declastruct context available to interfaces of this provider
     */
    context: DeclastructProviderContext<AO>,
  ) => Promise<Required<R>[]>;

  /**
   * a method capable of creating a resource of this class
   */
  create: (
    /**
     * the state we want to create the resource with
     */
    resource: R,

    /**
     * the declastruct context available to interfaces of this provider
     */
    context: DeclastructProviderContext<AO>,
  ) => Promise<Required<R>>;

  /**
   * a method capable of destroying a resource of this class
   */
  destroy: (
    /**
     * the primary key of the resource we want to destroy
     */
    primary: Required<Pick<R, P>>,

    /**
     * the declastruct context available to interfaces of this provider
     */
    context: DeclastructProviderContext<AO>,
  ) => Promise<Required<R>>;

  /**
   * a method capable of creating a resource of this class
   *
   * note
   * - may not be supported if resource is immutable
   */
  update?: (
    /**
     * the state we want to update the resource to
     *
     * note
     * - either primary or unique key must be defined
     * - any updatable keys defined will be persisted
     */
    resource: R,

    /**
     * the declastruct context available to interfaces of this provider
     */
    context: DeclastructProviderContext<AO>,
  ) => Promise<Required<R>>;
}
