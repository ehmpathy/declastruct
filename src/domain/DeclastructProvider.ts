/* eslint-disable @typescript-eslint/ban-types */
import type { LogMethods } from 'simple-leveled-log-methods';

import { DeclaredResource } from './DeclaredResource';
import { DeclastructProviderContext } from './DeclastructProviderContext';
import { DeclastructProviderResourceRemoteStateInterface } from './DeclastructProviderResourceRemoteStateInterface';

/**
 * the class name of a declared resource
 *
 * note
 * - it is just a string, but we distinguish it with its own type for clarity when defining other types
 */
export type DeclaredResourceClassName = string;

/**
 * agent options can take any shape
 *
 * typically, agent options include
 * - credentials
 * - caching mechanisms
 * - request throttling mechanisms
 */
export type DeclastructProviderAgentOptions = {
  log: LogMethods;
};

/**
 * a provider exposes the agent, interfaces, and resource definitions required to declaratively manage resources
 */
export interface DeclastructProvider<
  AO extends DeclastructProviderAgentOptions,
> {
  /**
   * defines the options available to agents of this provider
   *
   * relevance
   * - agents act on the behalf of the user to manage their resources
   * - typically, agent options include
   *   - credentials
   *   - caching mechanisms
   *   - request throttling mechanisms
   */
  agentOptions: AO;

  /**
   * defines the interfaces exposed by the provider for managing the remote state of resources
   *
   * note
   * - specified as a "name" to "interface" lookup table, for convenient usage
   */
  interfaces: Record<
    DeclaredResourceClassName,
    DeclastructProviderResourceRemoteStateInterface<
      DeclaredResource,
      DeclastructProviderContext<AO>,
      keyof DeclaredResource,
      keyof DeclaredResource
    >
  >;

  /**
   * defines hooks that we should execute in conjunction with managing resources
   */
  hooks: {
    /**
     * a hook to run before any operations for this provider
     */
    beforeAll?: () => {};

    /**
     * a hook to run after all operations for this provider
     */
    afterAll?: () => {};
  };
}
