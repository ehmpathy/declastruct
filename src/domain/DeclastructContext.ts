import type { LogMethods } from 'simple-leveled-log-methods';

import { DeclaredResource } from './DeclaredResource';
import { DeclastructProvider } from './DeclastructProvider';

/**
 * the declastruct context defines all of the providers and resources declared by the user
 *
 * note:
 * - this is the global context available to declastruct methods
 * - this is _not_ the provider context available to provider methods
 */
export interface DeclastructContext {
  /**
   * the providers loaded by the user
   */
  providers: DeclastructProvider<any>[];

  /**
   * all of the resources declared by the user
   */
  resources: DeclaredResource[];

  /**
   * the log context to use
   */
  log: LogMethods;
}
