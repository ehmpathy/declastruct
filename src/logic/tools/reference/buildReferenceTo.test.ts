import { DomainObject } from 'domain-objects';

import { DeclaredResourceReference } from '../../../domain/DeclaredResourceReference';
import {
  CanNotReferenceDeclaredResourceClassError,
  buildReferenceTo,
} from './buildReferenceTo';

// todo: remove the declastruct references in favor of domain-object references
describe.skip('getReferenceTo', () => {
  it('should throw a helpful error the class does not have unique key defined', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = ['uuid'];
    }
    type CNCMachineReference = DeclaredResourceReference<
      CNCMachine,
      'uuid',
      'serialNumber'
    >;
    try {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const ref: CNCMachineReference = buildReferenceTo(
        CNCMachine,
        new CNCMachine({
          serialNumber: '821',
          size: 'big',
          location: 'basement',
        }),
      );
      throw new Error('should not reach here');
    } catch (error: any) {
      expect(error).toBeInstanceOf(CanNotReferenceDeclaredResourceClassError);
      expect(error.message).toContain(
        `the static property 'unique' was not defined as an array of strings`,
      );
      expect(error).toMatchSnapshot();
    }
  });
  it('should throw a helpful error the class does not have primary key defined correctly', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = 'uuid';
    }
    type CNCMachineReference = DeclaredResourceReference<
      CNCMachine,
      'uuid',
      'serialNumber'
    >;
    try {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const ref: CNCMachineReference = buildReferenceTo(
        CNCMachine,
        new CNCMachine({
          serialNumber: '821',
          size: 'big',
          location: 'basement',
        }),
      );
      throw new Error('should not reach here');
    } catch (error: any) {
      expect(error).toBeInstanceOf(CanNotReferenceDeclaredResourceClassError);
      expect(error.message).toContain(
        `the static property 'primary' was not defined as an array of strings`,
      );
      expect(error).toMatchSnapshot();
    }
  });
  it('should be able to reference by unique key', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = ['uuid'];
      public static unique = ['serialNumber'];
    }
    type CNCMachineReference = DeclaredResourceReference<
      CNCMachine,
      'uuid',
      'serialNumber'
    >;
    const reference: CNCMachineReference = buildReferenceTo(
      CNCMachine,
      new CNCMachine({
        serialNumber: '821',
        size: 'big',
        location: 'basement',
      }),
    );
    expect(reference).toEqual({
      referenceOf: 'CNCMachine',
      identifiedBy: {
        key: 'UNIQUE_KEY',
        value: { serialNumber: '821' },
      },
    });
    expect(reference).toMatchSnapshot(); // log for viewing
  });
  it('should be able to reference by primary key', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = ['uuid'];
      public static unique = ['serialNumber'];
    }
    type CNCMachineReference = DeclaredResourceReference<
      CNCMachine,
      'uuid',
      'serialNumber'
    >;
    const reference: CNCMachineReference = buildReferenceTo(CNCMachine, {
      uuid: '__UUID__',
    });
    expect(reference).toEqual({
      referenceOf: 'CNCMachine',
      identifiedBy: {
        key: 'PRIMARY_KEY',
        value: { uuid: '__UUID__' },
      },
    });
    expect(reference).toMatchSnapshot(); // log for viewing
  });
  it('should reference by unique key if both are available', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = ['uuid'];
      public static unique = ['serialNumber'];
    }
    type CNCMachineReference = DeclaredResourceReference<
      CNCMachine,
      'uuid',
      'serialNumber'
    >;
    const reference: CNCMachineReference = buildReferenceTo(
      CNCMachine,
      new CNCMachine({
        uuid: 'de0e34ca-ce61-4884-ace5-093a1a32ff92',
        serialNumber: '821',
        size: 'big',
        location: 'basement',
      }),
    );
    expect(reference).toEqual({
      referenceOf: 'CNCMachine',
      identifiedBy: {
        key: 'UNIQUE_KEY',
        value: { serialNumber: '821' },
      },
    });
    expect(reference).toMatchSnapshot(); // log for viewing
  });
  it('should be able to instantiate with more specific reference class if one exists', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = ['uuid'];
      public static unique = ['serialNumber'];
    }
    class CNCMachineReference extends DeclaredResourceReference<
      CNCMachine,
      'uuid',
      'serialNumber'
    > {}
    const reference = buildReferenceTo(
      CNCMachine,
      new CNCMachine({
        uuid: 'de0e34ca-ce61-4884-ace5-093a1a32ff92',
        serialNumber: '821',
        size: 'big',
        location: 'basement',
      }),
      CNCMachineReference,
    );
    expect(reference).toBeInstanceOf(CNCMachineReference);
    expect(reference).toMatchSnapshot(); // log for viewing
  });
});
