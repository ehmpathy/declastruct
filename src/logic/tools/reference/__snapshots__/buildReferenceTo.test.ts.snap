// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReferenceTo should be able to instantiate with more specific reference class if one exists 1`] = `
CNCMachineReference {
  "identifiedBy": DeclaredResourceReferenceIdentifier {
    "key": "UNIQUE_KEY",
    "value": DeclaredResourceReferenceIdentifierValue {
      "serialNumber": "821",
    },
  },
  "referenceOf": "CNCMachine",
}
`;

exports[`getReferenceTo should be able to reference by primary key 1`] = `
DeclaredResourceReference {
  "identifiedBy": DeclaredResourceReferenceIdentifier {
    "key": "PRIMARY_KEY",
    "value": DeclaredResourceReferenceIdentifierValue {
      "uuid": "__UUID__",
    },
  },
  "referenceOf": "CNCMachine",
}
`;

exports[`getReferenceTo should be able to reference by unique key 1`] = `
DeclaredResourceReference {
  "identifiedBy": DeclaredResourceReferenceIdentifier {
    "key": "UNIQUE_KEY",
    "value": DeclaredResourceReferenceIdentifierValue {
      "serialNumber": "821",
    },
  },
  "referenceOf": "CNCMachine",
}
`;

exports[`getReferenceTo should reference by unique key if both are available 1`] = `
DeclaredResourceReference {
  "identifiedBy": DeclaredResourceReferenceIdentifier {
    "key": "UNIQUE_KEY",
    "value": DeclaredResourceReferenceIdentifierValue {
      "serialNumber": "821",
    },
  },
  "referenceOf": "CNCMachine",
}
`;

exports[`getReferenceTo should throw a helpful error the class does not have primary key defined correctly 1`] = `[Error: Can not reference declared resource of class 'CNCMachine'. Instances of the class 'CNCMachine' can not be referenced because the static property 'primary' was not defined as an array of strings.]`;

exports[`getReferenceTo should throw a helpful error the class does not have unique key defined 1`] = `[Error: Can not reference declared resource of class 'CNCMachine'. Instances of the class 'CNCMachine' can not be referenced because the static property 'unique' was not defined as an array of strings.]`;
